### PR TITLE
Add project sharing and permission propagation

### PR DESCRIPTION
**Pull request by arjunrajlab**
## Summary
This PR implements comprehensive project sharing functionality with automatic permission propagation to all contained resources (datasets, collections, and dataset views). Users with admin access to a project can now share it with other users, make it public/private, and manage access levels, with changes automatically cascading to all related resources.

### Backend API Endpoints (project.py)
- Added `POST /project/:id/share` - Share project with users at specific access levels (READ, WRITE, or revoke with -1)
- Added `GET /project/:id/access` - Retrieve current access list including users and public status
- Added `POST /project/:id/set_public` - Make project and all resources public or private
- Modified `addDataset()` and `addCollection()` to automatically sync permissions when resources are added to shared projects

### Permission Propagation (project.py model)
- `propagateUserAccess()` - Applies a user's access level to all project resources (datasets, collections, views)
- `propagatePublic()` - Sets public/private status on all project resources
- `propagateAccessToDataset()` - Syncs project ACL when a dataset is added
- `propagateAccessToCollection()` - Syncs project ACL when a collection is added
- `propagateAllUsersAccess()` - Helper to apply full project user list to new resources
- `_gatherAllResources()` - Bulk-loads all datasets, collections, and views linked to a project

### Bulk DB Operations (review feedback)
Permission propagation now uses MongoDB bulk write operations instead of per-document calls:

| Operation | Before | After |
|-----------|--------|-------|
| Share with 1 user (N resources) | 3N DB calls | 6 calls max |
| Remove user access (N resources) | 3N DB calls | 3 calls |
| Set public (N resources) | 3N DB calls | 3 calls |
| Add dataset to shared project (M users) | M × (1 + K) calls | 7 calls max |

New private helpers:
- `_bulkSetUserAccess()` - `$pull` + `$push` for single-user access across multiple docs
- `_bulkSetPublic()` - `$set` with `updateMany` for public flag
- `_bulkSetAllUsersAccess()` - Bulk user validation + `$pull`/`$push` with `$each` for full ACL sync

### Code Factorization (review feedback)
Extracted repeated patterns in the project model and API into reusable helpers:

| Pattern | Helper | Occurrences Consolidated |
|---------|--------|--------------------------|
| Access list formatting | `formatAccessList()` in `access_helpers.py` | 2 (project + datasetView APIs) |
| ObjectId conversion | `_ensureObjectId()` | 4 |
| Add/remove resource | `_addResource()` / `_removeResource()` | 4 methods |
| Apply to all resource types | `_applyToAllResourceTypes()` | 2 methods |
| Propagate to resource | `_propagateToResource()` | 4 methods |

Net result: ~120 fewer lines while preserving all public method signatures and behavior.

### Shared Helpers
- Extracted `fetchUserEmails()` into `access_helpers.py` (used by both project and datasetView endpoints)
- Extracted `formatAccessList()` into `access_helpers.py` (used by both project and datasetView endpoints)

### Frontend Components
- New `ShareProject.vue` component with UI for:
  - Public/private toggle
  - Current access list with user details
  - Add user form with access level selection
  - Remove user functionality with confirmation dialog
- Updated `ProjectInfo.vue` to include "Share Project" button
- Added API methods in `ProjectsAPI.ts` for share, getAccess, and setPublic operations
- Added `IProjectAccessList` interface to model types

### Testing
- Comprehensive test suite in `test_project.py` covering:
  - Permission propagation to datasets and collections
  - Access revocation cascading to resources
  - ACL sync when adding datasets to shared projects
  - Public/private propagation across all resources
  - **Bulk operation tests**: multi-resource sharing, multi-user ACL sync, access level upgrades, empty project edge case, full revoke across all resource types

### Skill Updates
- Updated nimbus-local-ops skill with curl parameter handling gotchas (autoDescribeRoute vs describeRoute, formData vs query params, shell variable expansion)

## Test plan
- [x] `flake8` passes on all changed files
- [x] `tox` - all 91 tests pass
- [x] Live curl tests: create project, add dataset, share with user, verify propagation, set public, verify propagation, cleanup
- [ ] Manual UI test: share project, verify user sees shared datasets

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py [hunk 1]
```diff
@@ -276,3 +278,340 @@ def updateStatus(self, project, status):
...
+        :param targetUser: The Girder user document.
+        :param level: AccessType level (0, 1, 2) or -1
+                      to remove access.
+        """
+        from ..models.datasetView import (
+            DatasetView as DatasetViewModel
+        )
+        from ..models.collection import (
+            Collection as CollectionModel
+        )
```
Comment (pchoisel) [inline devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py]: Why not have them both at the top of the file ?

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py [hunk 2]
```diff
@@ -276,3 +278,340 @@ def updateStatus(self, project, status):
...
+                ds, targetUser, level, save=True
+            )
+        for coll in resources['collections']:
+            CollectionModel().setUserAccess(
+                coll, targetUser, level, save=True
+            )
+        for dv in resources['datasetViews']:
+            DatasetViewModel().setUserAccess(
+                dv, targetUser, level, save=True
+            )
```
Comment (pchoisel) [inline devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py]: This will create a lot of requests on the database. Instead it would be better to do a batch update: `https://www.mongodb.com/docs/manual/core/bulk-write-operations/`  
However there is not batch setUserAccess in Girder that I'm aware of, so that would be something you'd have to implement.

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py [hunk 3]
```diff
@@ -276,3 +280,400 @@ def updateStatus(self, project, status):
...
+            [collection['_id']], CollectionModel(), True
+        )
+        dv_ids = [v['_id'] for v in dataset_views]
+        self._bulkSetPublic(
+            dv_ids, DatasetViewModel(), True
+        )
+        ds_ids = [d['_id'] for d in datasets]
+        self._bulkSetPublic(
+            ds_ids, Folder(), True
+        )
```
[the remark this answers is not in the document]
Comment (arjunrajlab) [reply devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/models/project.py]: Factorized!
